File: collect_app/src/main/java/org/odk/collect/android/widgets/DrawWidget.java
/*
 * Copyright (C) 2012 University of Washington
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

package org.odk.collect.android.widgets;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.view.View;
import android.widget.Button;
import org.javarosa.form.api.FormEntryPrompt;
import org.odk.collect.android.R;
import org.odk.collect.android.activities.DrawActivity;

import static org.odk.collect.android.utilities.ApplicationConstants.RequestCodes;

/**
 * Free drawing widget.
 *
 * @author BehrAtherton@gmail.com
 */
@SuppressLint("ViewConstructor")
public class DrawWidget extends BaseImageWidget {

    private Button drawButton;

    public DrawWidget(Context context, FormEntryPrompt prompt) {
        super(context, prompt);
        imageClickHandler = new DrawImageClickHandler(DrawActivity.OPTION_DRAW, RequestCodes.DRAW_IMAGE, R.string.draw_image);
        setUpLayout();
        setUpBinary();
        addAnswerView(answerLayout);
    }

    @Override
    protected void setUpLayout() {
        super.setUpLayout();
        drawButton = getSimpleButton(getContext().getString(R.string.draw_image));
        drawButton.setEnabled(!getFormEntryPrompt().isReadOnly());

        answerLayout.addView(drawButton);
        answerLayout.addView(errorTextView);

        if (getFormEntryPrompt().isReadOnly()) {
            drawButton.setVisibility(View.GONE);
        }
        errorTextView.setVisibility(View.GONE);
    }

    @Override
    public Intent addExtrasToIntent(Intent intent) {
        return intent;
    }

    @Override
    public void clearAnswer() {
        super.clearAnswer();
        // reset buttons
        drawButton.setText(getContext().getString(R.string.draw_image));
    }

    @Override
    public void setOnLongClickListener(OnLongClickListener l) {
        drawButton.setOnLongClickListener(l);
        super.setOnLongClickListener(l);
    }

    @Override
    public void cancelLongPress() {
        super.cancelLongPress();
        drawButton.cancelLongPress();
    }

    @Override
    public void onButtonClick(int buttonId) {
        imageClickHandler.clickImage("drawButton");
    }
}

File: collect_app/src/main/java/org/odk/collect/android/widgets/DrawWidget.kt
package org.odk.collect.android.widgets

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.view.View
import org.javarosa.form.api.FormEntryPrompt
import org.odk.collect.android.databinding.DrawWidgetBinding
import org.odk.collect.android.formentry.questions.QuestionDetails
import org.odk.collect.android.utilities.ApplicationConstants.RequestCodes
import org.odk.collect.android.utilities.QuestionMediaManager
import org.odk.collect.android.widgets.utilities.WaitingForDataRegistry
import org.odk.collect.draw.DrawActivity
import org.odk.collect.strings.R

@SuppressLint("ViewConstructor")
class DrawWidget(
    context: Context,
    prompt: QuestionDetails,
    questionMediaManager: QuestionMediaManager,
    waitingForDataRegistry: WaitingForDataRegistry,
    tmpImageFilePath: String
) : BaseImageWidget(
        context,
        prompt,
        questionMediaManager,
        waitingForDataRegistry,
        tmpImageFilePath
    ) {
    lateinit var binding: DrawWidgetBinding

    init {
        imageClickHandler = DrawImageClickHandler(
            DrawActivity.OPTION_DRAW,
            RequestCodes.DRAW_IMAGE,
            R.string.draw_image
        )

        render()
        updateAnswer()
    }

    override fun onCreateAnswerView(context: Context, prompt: FormEntryPrompt, answerFontSize: Int): View {
        binding = DrawWidgetBinding.inflate((context as Activity).layoutInflater)
        binding.drawButton.setOnClickListener {
            imageClickHandler.clickImage(
                "drawButton"
            )
        }
        binding.answerView.setOnClickListener { imageClickHandler.clickImage("viewImage") }

        if (questionDetails.isReadOnly) {
            binding.drawButton.visibility = GONE
        }

        answerView = binding.answerView
        answerView.setup(prompt.answerValue, imageLoader, questionMediaManager, referenceManager, getDefaultFilePath())
        return binding.root
    }

    override fun addExtrasToIntent(intent: Intent) = intent

    override fun clearAnswer() {
        super.clearAnswer()
        binding.drawButton.text = context.getString(R.string.draw_image)
    }

    override fun setOnLongClickListener(l: OnLongClickListener?) {
        binding.drawButton.setOnLongClickListener(l)
        super.setOnLongClickListener(l)
    }

    override fun cancelLongPress() {
        super.cancelLongPress()
        binding.drawButton.cancelLongPress()
    }
}
